demos/common/monitors/cpu_monitor.cpp

```diff
@@ -0,0 +1,274 @@
+    }
+    else {
+        this->historySize = historySize;
+        historyEnabled = true;
+        if (!lastEnabled) {
+            openQuery();
+        }
+    }
+}
+
+bool CpuMonitor::isHistoryEnabled() const {
+    return historyEnabled;
+}
+
+void CpuMonitor::disableHistory() {
+    historyEnabled = false;
+    if (!lastEnabled) {
+        closeQuery();
+    }
+}
+
+void CpuMonitor::enableLast() {
+    lastEnabled = true;
+    if (!historyEnabled) {
+        historySize = 1;
+        openQuery();
+    }
+}
+
+bool CpuMonitor::isLastEnabled() const {
+    return lastEnabled;
+}
+
+void CpuMonitor::disableLast() {
+    lastEnabled = false;
+    if (!historyEnabled) {
+        closeQuery();
+    }
+}
+
+void CpuMonitor::collectData() {
+    // store CPU Utulization in range [0, 1]
+    PDH_STATUS status;
+    PDH_FMT_COUNTERVALUE DisplayValue;
+
+    status = PdhCollectQueryData(query);
+    if (ERROR_SUCCESS != status) {
+        throw std::runtime_error("PdhCollectQueryData() failed");
+    }
+
+    std::vector<std::pair<double, double>> idleNonIdleCpuStat;
+    for (std::size_t i = 0; i < coreTimeCounters.size(); i++) {
+        status = PdhGetFormattedCounterValue(coreTimeCounters[i], PDH_FMT_DOUBLE, NULL, &DisplayValue);
+        if (ERROR_SUCCESS != status) {
+            throw std::runtime_error("PdhGetFormattedCounterValue() failed");
+        }
+
+        idleNonIdleCpuStat.emplace_back(DisplayValue.doubleValue, 1 - DisplayValue.doubleValue);
+    }
+
+    std::vector<double> cpuLoad(idleNonIdleCpuStat.size());
+    for (std::size_t i = 0; i < idleNonIdleCpuStat.size(); ++i) {
+        cpuLoad[i] = static_cast<double>(idleNonIdleCpuStat[i].first);
+    }
+
+    if (meanCpuLoad.empty()) {
+        meanCpuLoad = cpuLoad;
+        samplesNumber = 1;
+    } else {
+        for (std::size_t i = 0; i < meanCpuLoad.size(); ++i) {
+            meanCpuLoad[i] = (meanCpuLoad[i] * samplesNumber + cpuLoad[i]) / (samplesNumber + 1);
+        }
+        ++samplesNumber;
+    }
+    cpuLoadHistory.push_back(std::move(cpuLoad));
+    if (cpuLoadHistory.size() > historySize) {
+        cpuLoadHistory.pop_front();
+    }
+}
+
+std::deque<std::vector<double>> CpuMonitor::getLastHistory() const {
+    return cpuLoadHistory;
+}
+
+std::vector<double> CpuMonitor::getMeanCpuLoad() const {
+    return meanCpuLoad;
+}
+#else
+#include <regex>
+#include <utility>
+#include <fstream>
+
+namespace {
+std::size_t getNCores() {
+    std::ifstream cpuinfo("/proc/cpuinfo");
+    return std::count(std::istream_iterator<std::string>(cpuinfo),
+        std::istream_iterator<std::string>(),
+        std::string("processor"));
+}
+
+std::vector<std::pair<unsigned long, unsigned long>> getIdleNonIdleCpuStat(std::size_t nCores) {
+    std::vector<std::pair<unsigned long, unsigned long>> idleNonIdleCpuStat(nCores);
+    std::ifstream proc_stat("/proc/stat");
+    std::string line;
+    std::smatch match;
+    std::getline(proc_stat, line);
+    while (proc_stat.good())
```

Comment:
A simpler way to iterate over lines is `while (std::getline(proc_stat, line))`.

Reply:
updated

Reply:
There are also a few other instances of such loops.

Reply:
updated